src/modules/Playlists/XSPF.cpp

```diff
@@ -0,0 +1,176 @@
+	Entry curr;
+	while (!xmlReader.atEnd() && !xmlReader.hasError())
+	{
+		QXmlStreamReader::TokenType token = xmlReader.readNext();
+		if (token == QXmlStreamReader::StartDocument)
+			continue;
+		if (token == QXmlStreamReader::StartElement)
+		{
+			const QStringRef name = xmlReader.name();
+			if (name == "track")
+			{
+				if (inEntry)
+				{
+					if (curr.name.isEmpty())
+						curr.name = Functions::fileName(curr.url, false);
+					list += curr;
+				}
+				curr = Entry();
+				inEntry = true;
+			}
+			else if (inEntry)
+			{
+				const QString value = xmlReader.readElementText();
+				if (name == "location")
+				{
+					if (value.startsWith("file://")) // local absolute path
+						curr.url = Functions::Url(value);
+					else if (Functions::splitPrefixAndUrlIfHasPluginPrefix(value, NULL, NULL, NULL)) // plugin URL
+						curr.url = value;
+					else // relative path
+						foreach (const QString &dir, m_relativeDirs)
+							if (QFile(dir + "/" + value).exists())
+								curr.url = Functions::Url(value, dir);
+				}
+				else if (name == "title")
+					(curr.name = value).replace('\001', '\n');
+				else if (name == "duration" && curr.length == -1.0)
+					curr.length = value.toInt() / 1000.0;
+				else if (name == "QMPlay_length")
+					curr.length = value.toDouble();
+				else if (name == "QMPlay_sel")
+					curr.selected = value.toInt();
+				else if (name == "QMPlay_queue")
+					curr.queue = value.toInt();
+				else if (name == "QMPlay_GID")
+					curr.GID = value.toInt();
+				else if (name == "QMPlay_parent")
+					curr.parent = value.toInt();
+			}
+		}
+		else if (token == QXmlStreamReader::EndElement && xmlReader.name() == "track")
+		{
+			if (inEntry)
+			{
+				if (curr.name.isEmpty())
+					curr.name = Functions::fileName(curr.url, false);
+				list += curr;
+			}
+			curr = Entry();
+			inEntry = false;
+		}
+	}
+	return list;
+}
+bool XSPF::write(const Entries &list)
+{
+	QByteArray buffer;
+	QXmlStreamWriter xmlWriter(&buffer);
+	xmlWriter.setAutoFormatting(true);
+	xmlWriter.setAutoFormattingIndent(-1);
+
+	xmlWriter.writeStartDocument();
+	xmlWriter.writeStartElement("playlist");
+	xmlWriter.writeStartElement("trackList");
+
+	foreach (const Entry &entry, list)
+	{
+		xmlWriter.writeStartElement("track");
+
+		if (!entry.url.isEmpty())
+		{
+			QString url = entry.url;
+			if (m_useRelative && url.startsWith("file://"))
+			{
+				QString path = url;
+				path.remove(0, 7);
+#ifdef Q_OS_WIN
+				if (path.startsWith('/'))
+					path.remove(0, 1);
+#endif
+				foreach (const QString &dir, m_relativeDirs)
+				{
+					if (path.startsWith(dir))
+					{
+						url = path.remove(0, dir.length() + (dir.endsWith('/') ? 0 : 1));
+						break;
+					}
+				}
+			}
+			xmlWriter.writeTextElement("location", url);
+		}
+		if (!entry.name.isEmpty())
+			xmlWriter.writeTextElement("title", entry.name);
+		if (entry.length >= 0.0)
+		{
+			xmlWriter.writeTextElement("duration", QString::number((qint32)(1000 * entry.length)));
```

Comment:
`duration` is in milliseconds, so it will be enough. `QMPlay_length` is not necessary in this case. Use `duration` and divide it by `1000.0` in read method.